presentation/components/src/presentation-components/table/DataProvider.ts

```diff
@@ -408,24 +408,52 @@ const createRow = (descriptor: Readonly<Descriptor>, item: Readonly<Item>, sameI
     };
   }
 
-  return {
-    key,
-    cells: updatedFields.map((field: Field): CellItem => {
-      const itemProps: Partial<CellItem> = {};
-      const mergedCellsCount = mergedCellsCounts[field.name];
-      if (mergedCellsCount) {
-        itemProps.mergedCellsCount = mergedCellsCount;
-        itemProps.alignment = HorizontalAlignment.Center;
-
-        const nestedField = sameInstanceFieldsMap[field.name];
-        nestedField.name = field.name;
-        field = nestedField;
-      }
-      return {
-        key: field.name,
-        record: ContentBuilder.createPropertyRecord({ field }, updatedItem).record,
-        ...itemProps,
-      };
-    }),
-  };
+  const builder = new CellsBuilder(mergedCellsCounts, sameInstanceFieldsMap);
+  traverseContentItem(builder, { ...descriptor, fields: updatedFields }, updatedItem);
+  return { key, cells: builder.cells };
 };
+
+class CellsBuilder extends PropertyRecordsBuilder {
+  private _cells?: CellItem[];
+
+  public constructor(
+    private _mergedCellCounts: { [fieldName: string]: number },
+    private _sameInstanceFields: { [fieldName: string]: Field },
+  ) {
+    super();
+  }
+
+  public get cells(): CellItem[] {
+    assert(this._cells !== undefined);
+    return this._cells;
+  }
+
+  protected createRootPropertiesAppender() {
+    return {
+      append: (record: FieldHierarchyRecord) => {
+        assert(this._cells !== undefined);
+        const itemProps: Partial<CellItem> = {};
+        const mergedCellsCount = this._mergedCellCounts[record.fieldHierarchy.field.name];
+        if (mergedCellsCount) {
+          itemProps.mergedCellsCount = mergedCellsCount;
+          itemProps.alignment = HorizontalAlignment.Center;
+
+          // FIXME: what is this for? all tests pass with it commented-out
+          // const nestedField = this._sameInstanceFields[record.fieldHierarchy.field.name];
+          // nestedField.name = record.fieldHierarchy.field.name;
+          // record.fieldHierarchy.field = nestedField;

```

Comment:
@aurislt7 can you remind me what this is for? None of the tests seem to be affected by this piece of code.

Reply:
Sure, this piece of code is responsible for mapping the nestedField information (without any modifications) to the field with the mergedCellsCount specified (used when there's more than one item in the nestedField). If these lines are removed, nestedField's information would be lost. It's unfortunate that tests don't get affected by it, however it should be easy to fix it by adding an assertion that checks the value of the merged cell.

Reply:
I improved on the full stack tests to capture this and fixed the data provider. As discussed on call, there are some issues not related to changes in this PR which we'll need to fix afterwards.